File: app/src/main/java/com/luki/play/bridge/BridgeMessage.kt
// bridge/BridgeMessage.kt
package com.luki.play.bridge

import org.json.JSONObject

/**
 * Sealed hierarchy for every message the web layer can send to the
 * native bridge via `window.LukiNative.<method>(jsonString)`.
 *
 * All messages share a [type] discriminator so that [LukiBridge] can
 * dispatch them in a single entry-point if needed, or the web can call
 * each method directly (both patterns are supported).
 */
sealed class BridgeMessage {

    // ------------------------------------------------------------------ //
    //  Playback
    // ------------------------------------------------------------------ //

    /**
     * Web requests native HLS playback.
     *
     * JSON shape:
     * ```json
     * {
     *   "url":              "http://...",
     *   "title":            "Canal HD",
     *   "poster":           "http://...",
     *   "subtitleUri":      "http://.../subs.vtt",
     *   "subtitleMimeType": "text/vtt"
     * }
     * ```
     */
    data class PlayStream(
        val url: String,
        val title: String,
        val poster: String?,
        val subtitleUri: String?,
        val subtitleMimeType: String?
    ) : BridgeMessage()

    /** Web requests the player to stop / return from PlayerActivity. */
    object StopStream : BridgeMessage()

    // ------------------------------------------------------------------ //
    //  Session / Auth
    // ------------------------------------------------------------------ //

    /**
     * Web signals that the user has completed login.
     * Native side can react (e.g. hide progress indicators, store tokens).
     *
     * JSON shape:
     * ```json
     * { "userId": "42", "displayName": "Marco" }
     * ```
     */
    data class LoginSuccess(val userId: String, val displayName: String) : BridgeMessage()

    /** Web signals that the user has logged out; native clears any cached state. */
    object Logout : BridgeMessage()

    // ------------------------------------------------------------------ //
    //  Device / UI
    // ------------------------------------------------------------------ //

    /**
     * Web requests entering Picture-in-Picture (mobile only).
     * Native checks [DeviceUtilsContract.supportsPip] before acting.
     */
    object EnterPip : BridgeMessage()

    /**
     * Web requests the current device context so it can adjust its layout.
     * Native responds via [LukiBridge.getDeviceInfo].
     */
    object GetDeviceInfo : BridgeMessage()

    // ------------------------------------------------------------------ //
    //  Parse helper
    // ------------------------------------------------------------------ //

    companion object {
        /**
         * Parses a raw JSON string arriving from the WebView JS layer.
         * Returns `null` for unrecognised [type] values (logged, not thrown).
         */
        fun from(json: String): BridgeMessage? {
            return try {
                val obj  = JSONObject(json)
                val type = obj.optString("type")
                when (type) {
                    "play_stream"   -> PlayStream(
                        url              = obj.getString("url"),
                        title            = obj.optString("title", ""),
                        poster           = obj.optString("poster").ifBlank { null },
                        subtitleUri      = obj.optString("subtitleUri").ifBlank { null },
                        subtitleMimeType = obj.optString("subtitleMimeType").ifBlank { null }
                    )
                    "stop_stream"   -> StopStream
                    "login_success" -> LoginSuccess(
                        userId      = obj.getString("userId"),
                        displayName = obj.optString("displayName", "")
                    )
                    "logout"        -> Logout
                    "enter_pip"     -> EnterPip
                    "get_device_info" -> GetDeviceInfo
                    else -> null
                }
            } catch (e: Exception) {
                null
            }
        }
    }
}

File: ios/LukiPlay/Core/Bridge/BridgeMessage.swift
import Foundation

enum BridgeMessage {
    case playStream(StreamConfig)
    case stopStream
    case loginSuccess(accessToken: String?, refreshToken: String?, userId: String?, displayName: String?)
    case logout
    case enterPip
    case getDeviceInfo(callbackId: String)
    case getStoredSession(callbackId: String)

    static func from(_ payload: Any) -> BridgeMessage? {
        let object: [String: Any]?

        if let string = payload as? String {
            let trimmed = string.trimmingCharacters(in: .whitespacesAndNewlines)
            if trimmed.lowercased() == "stop_stream" {
                return .stopStream
            }
            if trimmed.lowercased().hasPrefix("http://") || trimmed.lowercased().hasPrefix("https://") {
                return .playStream(StreamConfig(url: trimmed))
            }
        }

        if let dictionary = payload as? [String: Any] {
            object = dictionary
        } else if let string = payload as? String,
                  let data = string.data(using: .utf8),
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            object = json
        } else {
            object = nil
        }

        guard let object else { return nil }

        let type = (object["type"] as? String)?.lowercased() ?? ""

        if type.isEmpty, let url = object["url"] as? String, !url.isEmpty {
            let config = StreamConfig(
                url: url,
                title: object["title"] as? String ?? "",
                posterUrl: object["poster"] as? String,
                subtitleUri: object["subtitleUri"] as? String,
                subtitleMimeType: object["subtitleMimeType"] as? String ?? "text/vtt",
                drmToken: object["drmToken"] as? String
            )
            return .playStream(config)
        }

        switch type {
        case "play_stream":
            guard let url = object["url"] as? String, !url.isEmpty else { return nil }
            let config = StreamConfig(
                url: url,
                title: object["title"] as? String ?? "",
                posterUrl: object["poster"] as? String,
                subtitleUri: object["subtitleUri"] as? String,
                subtitleMimeType: object["subtitleMimeType"] as? String ?? "text/vtt",
                drmToken: object["drmToken"] as? String
            )
            return .playStream(config)

        case "stop_stream":
            return .stopStream

        case "login_success":
            return .loginSuccess(
                accessToken: object["accessToken"] as? String,
                refreshToken: object["refreshToken"] as? String,
                userId: object["userId"] as? String,
                displayName: object["displayName"] as? String
            )

        case "logout":
            return .logout

        case "enter_pip":
            return .enterPip

        case "get_device_info":
            return .getDeviceInfo(callbackId: object["callbackId"] as? String ?? "")

        case "get_stored_session":
            return .getStoredSession(callbackId: object["callbackId"] as? String ?? "")

        default:
            return nil
        }
    }
}
